doc/architecture-decisions/0012-stopped-app-routing.md

```diff
@@ -0,0 +1,66 @@
+# 12. Stopped App Routing
+
+Date: 2020-05-29
+
+## Status
+
+Proposal
+
+## Context
+
+In CF-for-VMs, when a route is mapped to multiple apps, some started and some
+stopped, gorouter only directs requests to the apps that are started.
+
+In CF-for-k8s _currently_, we balance traffic equally between the started and
+stopped apps. This results in 503 errors for the user, and is not feature
+pairity.
+
+This is because the components that create the Istio configuration do not
+currently know about the status of apps. They assume all destinations on a Route
+object are up and able to receive traffic.
+
+See this issue: https://github.com/cloudfoundry/cf-for-k8s/issues/158
+
+## Proposal
+
+We have two proposals:
+
+1. Cloud Controller only adds destinations to Routes that are running. This way
+   Route objects remain the single source of truth for routing in the platform.
+   * _Note:_ Routes with _no_ destinations are explicitly supported already.
+
+2. Cloud Controller writes an App CR (this is a new CRD) that contains the
+   application's status, so that the routecontroller can use that to decide how
+   to configure Istio.
+
+## Alternatives Considered
```

Comment:
What about using Kubernetes [Readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/#define-readiness-probes)?

Reply:
When an app is stopped in CF, the pod and statefulset are deleted, so we can't probe them for readiness.

Services which have selectors which are not fulfilled by any pods result in 503s. We have one service per CF app. So when an app is stopped the service for that app results in a 503.

One alternative we thought about that takes advantage of that aspect of services, but thought was too ridiculous to include was:
* Have capi label every pod with every route that maps to it
* Make one service per route that selects that label
* Make one virtualservice per route that selects the service

This doesn't solve for the situation when _no apps are started_, but it would prevent routing to apps that are down. It doesn't strike me as healthy to include routing information in the labels of the pods though, and it seemed like it'd be a lot of work for capi. You also lose weighted routing (and possibly some other features, haven't really checked) with this approach.

Reply:
If the app is stopped, there will be no eirini statefulset and therefore no pods to put readiness probes on.

Reply:
Makes sense.  May be worth summarizing this point as an "Alternative considered" so that future team members don't ask the same question that I did.